File: Golang/contiguousArray.go
/*
Given a binary array, find the maximum length of a
contiguous subarray with equal number of 0 and 1.
*/

package main

func findMaxLength(nums []int) int {
	count := 0
	maxL := 0

	myMap := make(map[int]int)
	myMap[0] = -1

	for i := 0; i < len(nums); i++ {
		if nums[i] == 1 {
			count++
		} else {
			count--
		}

		if idx, ok := myMap[count]; !ok {
			myMap[count] = i
		} else {
			maxL = max(maxL, i-idx)
		}
	}
	return maxL
}

File: Cplusplus/contiguousArray.cpp
// Given a binary array, find the maximum length of a contiguous subarray with equal number of 0 and 1.

#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

int findMaxLength(vector<int>& nums);

int main(int argc, char const *argv[])
{
    vector<int> nums = {0,1,0,0,1,1,0};
    cout << findMaxLength(nums) << endl;
    return 0;
}

int findMaxLength(vector<int>& nums)
{
    int count = 0;
    int maxL = 0;
    unordered_map<int, int> map;

    for (int i = 0; i < nums.size(); i++)
    {
        count = count + nums[i] == 0 ? -1 : 1;

        if (map.find(count) == map.end())
        {
            map[count] = i;
        }
        else
        {
            maxL = max(maxL, i - map[count]);
        }
    }
    return maxL;
}